File: tests/domains/video/test_concat.py
import pytest

from domains import Runtime
from domains.video.concat import concat_videos, transition_offsets
from domains.video.cut import cut_video
from domains.video.probe import probe_video
from tests.helpers import unwrap

pytestmark = pytest.mark.ffmpeg


def test_transition_offsets() -> None:
    assert transition_offsets([3.0, 3.0, 3.0], 0.5) == [2.5, 5.0]


@pytest.fixture
def clips(runtime: Runtime, sample_video: str) -> list[str]:
    first = unwrap(cut_video(runtime, sample_video, 0.0, 1.5, reencode=True))
    second = unwrap(cut_video(runtime, sample_video, 1.5, 3.0, reencode=True))
    return [first["output"], second["output"]]


def test_concat_copy(runtime: Runtime, clips: list[str]) -> None:
    result = unwrap(concat_videos(runtime, clips))
    assert result["count"] == 2
    assert result["transition"] is None
    assert result["has_audio"] is True
    probed = unwrap(probe_video(runtime, result["output"]))
    assert abs(probed["info"]["duration"] - 3.0) < 0.3


def test_concat_with_transition(runtime: Runtime, clips: list[str]) -> None:
    result = unwrap(
        concat_videos(
            runtime, clips, transition="fade", transition_duration=0.5, output_name="f.mp4"
        )
    )
    assert result["output"] == "f.mp4"
    assert result["transition"] == "fade"
    assert result["has_audio"] is True
    probed = unwrap(probe_video(runtime, result["output"]))
    assert abs(probed["info"]["duration"] - 2.5) < 0.3


def test_concat_transition_invalid(runtime: Runtime, clips: list[str]) -> None:
    assert concat_videos(runtime, clips, transition="explode").get("code") == (  # type: ignore[arg-type]
        "invalid_argument"
    )
    assert concat_videos(runtime, clips, transition="fade", transition_duration=2.0).get(
        "code"
    ) == ("invalid_argument")
    assert concat_videos(runtime, clips[:1]).get("code") == "invalid_argument"


def test_concat_missing(runtime: Runtime, sample_video: str) -> None:
    assert concat_videos(runtime, [sample_video, "ghost.mp4"]).get("code") == "not_found"

File: tests/domains/video/test_concat.rs
use mcp_tools::core::errors::ErrorCode;
use mcp_tools::domains::video::concat::{concat_videos, transition_offsets, TransitionName};
use mcp_tools::domains::video::cut::cut_video;
use mcp_tools::domains::video::probe::probe_video;

use crate::conftest::{runtime, sample_video, TestRuntime};
use crate::helpers::unwrap;
use crate::skip_without_ffmpeg;

#[test]
fn test_transition_offsets() {
    assert_eq!(transition_offsets(&[3.0, 3.0, 3.0], 0.5), vec![2.5, 5.0]);
}

fn clips(rt: &TestRuntime) -> Vec<String> {
    let sample = sample_video(&rt.runtime.workspace);
    let first = unwrap(cut_video(&rt.runtime, &sample, 0.0, 1.5, true, false));
    let second = unwrap(cut_video(&rt.runtime, &sample, 1.5, 3.0, true, false));
    vec![first.output, second.output]
}

#[test]
fn test_concat_copy() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let clips = clips(&rt);
    let result = unwrap(concat_videos(&rt.runtime, &clips, None, None, 0.5, false));
    assert_eq!(result.count, 2);
    assert!(result.transition.is_none());
    assert!(result.has_audio);
    let probed = probe_video(&rt.runtime, &result.output).unwrap();
    assert!((probed.info.duration - 3.0).abs() < 0.3);
}

#[test]
fn test_concat_with_transition() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let clips = clips(&rt);
    let result = unwrap(concat_videos(
        &rt.runtime,
        &clips,
        Some("f.mp4"),
        Some(TransitionName::Fade),
        0.5,
        false,
    ));
    assert_eq!(result.output, "f.mp4");
    assert_eq!(result.transition, Some(TransitionName::Fade));
    assert!(result.has_audio);
    let probed = probe_video(&rt.runtime, &result.output).unwrap();
    assert!((probed.info.duration - 2.5).abs() < 0.3);
}

#[test]
fn test_concat_transition_invalid() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let clips = clips(&rt);
    let error = concat_videos(
        &rt.runtime,
        &clips,
        None,
        Some(TransitionName::Fade),
        2.0,
        false,
    )
    .unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
    let error = concat_videos(&rt.runtime, &clips[..1], None, None, 0.5, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
}

#[test]
fn test_concat_missing() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let paths = vec![sample, "ghost.mp4".to_string()];
    let error = concat_videos(&rt.runtime, &paths, None, None, 0.5, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}
